### PR TITLE
Trace/db traces

### LINKED ISSUE #4205: Support tracing of synced, not-processed blocks
**Describe the bug**
I've synced Nethermind on the mainnet to replace OpenEthereum client, but trace_filter returns empty response for some reason:

```
curl --data '{"method":"trace_filter","params":[{"fromBlock":"0x1","toBlock":"0x50"}],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
{"jsonrpc":"2.0","result":[],"id":1}
```

On my old OpenEthereum node the result is correct though:
```
curl --data '{"method":"trace_filter","params":[{"fromBlock":"0x1","toBlock":"0x50"}],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST 127.0.0.1:9110
{"jsonrpc":"2.0","result":[{"action":{"author":"0x05a56e2d52c817161883f50c441c3228cfe54d9f","rewardType":"block","value":"0x4563918244f40000"},"blockHash":"0x88e96d4537bea4d9c05d12549907b32561d3bf31f45aae734cdc119f13406cb6","blockNumber":1,"result":null,"subtraces":0,"traceAddress":...":[],"transactionHash":null,"transactionPosition":null,"type":"reward"}],"id":1}
```

I've tried enumerating through latest blocks and found out that the only available blocks for trace_filter are last 2910 +/- 10 blocks from the latest synced block.

E.g. If I try to `trace_filter` block beyound that range `latest_block` - 3000 nethermind gives me an error:

```latest_block=$(echo "obase=16;$(~nethermind/blocknumber.sh)-3000"|bc);curl -s --data '{"method":"trace_filter","params":[{"fromBlock":"0x'$latest_block'","toBlock":"0x'$latest_block'"}],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545|json_pp
{
   "error" : {
      "code" : -32603,
      "data" : "Nethermind.Trie.TrieException: Failed to load key 17e18854d6c1cee0ac2cee66833b0c4ff4c250224b30519d295e3e503d9f5bab from root hash 0x16492ad5cecdae541bccf891c2e8ce664bb4ccc5be2cd74e7ce806d68302942b.\n ---> Nethermind.Trie.TrieException: Node 0x16492ad5cecdae541bccf891c2e8ce664bb4ccc5be2cd74e7ce806d68302942b is missing from the DB\n   at Nethermind.Trie.Pruning.TrieStore.LoadRlp(Keccak keccak, IKeyValueStore keyValueStore)\n   at Nethermind.Trie.TrieNode.ResolveNode(ITrieNodeResolver tree)\n   at Nethermind.Trie.PatriciaTree.Run(Span`1 updatePath, Int32 nibblesCount, Byte[] updateValue, Boolean isUpdate, Boolean ignoreMissingDelete, Keccak startRootHash)\n   at Nethermind.Trie.PatriciaTree.Get(Span`1 rawKey, Keccak rootHash)\n   --- End of inner exception stack trace ---\n   at Nethermind.Trie.PatriciaTree.Get(Span`1 rawKey, Keccak rootHash)\n   at Nethermind.State.StateProvider.GetState(Address address)\n   at Nethermind.State.StateProvider.GetAndAddToCache(Address address)\n   at Nethermind.State.StateProvider.GetNonce(Address address)\n   at Nethermind.Consensus.Processing.BlockProcessor.BlockValidationTransactionsExecutor.ProcessTransactions(Block block, ProcessingOptions processingOptions, BlockReceiptsTracer receiptsTracer, IReleaseSpec spec)\n   at Nethermind.Consensus.Processing.BlockProcessor.ProcessBlock(Block block, IBlockTracer blockTracer, ProcessingOptions options)\n   at Nethermind.Consensus.Processing.BlockProcessor.ProcessOne(Block suggestedBlock, ProcessingOptions options, IBlockTracer blockTracer)\n   at Nethermind.Consensus.Processing.BlockProcessor.Process(Keccak newBranchStateRoot, List`1 suggestedBlocks, ProcessingOptions options, IBlockTracer blockTracer)\n   at Nethermind.Consensus.Processing.BlockchainProcessor.ProcessBranch(ProcessingBranch processingBranch, Block suggestedBlock, ProcessingOptions options, IBlockTracer tracer)\n   at Nethermind.Consensus.Processing.BlockchainProcessor.Process(Block suggestedBlock, ProcessingOptions options, IBlockTracer tracer)\n   at Nethermind.Consensus.Processing.OneTimeChainProcessor.Process(Block block, ProcessingOptions options, IBlockTracer tracer)\n   at Nethermind.Consensus.Tracing.Tracer.Trace(Block block, IBlockTracer blockTracer)\n   at Nethermind.JsonRpc.Modules.Trace.TraceRpcModule.TraceBlock(Block block, ParityTraceTypes traceTypes, TxTraceFilter txTraceFilter)\n   at Nethermind.JsonRpc.Modules.Trace.TraceRpcModule.trace_filter(TraceFilterForRpc traceFilterForRpc)",
      "message" : "Internal error"
   },
   "jsonrpc" : "2.0",
   "id" : 1
}
```

NOTE: old OpenEthereum node was running with fast sync (NOT archive!) and `trace_filter` worked just fine there and consumed ~1.1T, new Nethermind node consumes little bit less ~729G but tracing seems to be not working. I do not want to run archive node, simply because of the storage requirements.

**To Reproduce**
1. Fully sync mainnet node
2. Use this configuration:
```
{
  "Init": {
    "ChainSpecPath": "chainspec/foundation.json",
    "GenesisHash": "0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3",
    "BaseDbPath": "nethermind_db/mainnet",
    "LogFileName": "mainnet.log",
    "MemoryHint": 10480000000
  },
  "Network": {
    "ActivePeersMaxCount": 100
  },
  "Sync": {
    "FastSync": true,
    "SnapSync": true,
    "PivotNumber": 14957000,
    "PivotHash": "0x7ce2c9cb601721e33ee6313e91b30c9f035b74ee282402b5bdfb1095c572acd7",
    "PivotTotalDifficulty": "51613742220579521774353",
    "FastBlocks": true,
    "WitnessProtocolEnabled": true
  },
  "JsonRpc": {
    "Enabled": true,
    "Timeout": 20000,
    "Host": "0.0.0.0",
    "Port": 8545,
    "EnabledModules": [
      "Eth",
      "Subscribe",
      "Trace",
      "TxPool",
      "Web3",
      "Personal",
      "Proof",
      "Net",
      "Parity",
      "Health",
      "Rpc",
    ],
    "AdditionalRpcUrls": [
      "http://127.0.0.1:8551|http;ws|net;eth;subscribe;engine;web3;client;rpc|"
    ],
    "IpcUnixDomainSocketPath":"/home/nethermind/.nethermind/nethermind.ipc"
  },
  "Metrics": {
    "NodeName": "Mainnet"
  }
}
```

3. Try to query the node with the above curl 

**Expected behavior**
Nethermind should return trace filter info same as OpenEthereum

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Version: `1.13.3-14-g2500d3daf` (built from master branch)

### PR DESCRIPTION
**Pull request by LukaszRozmej**
Resolves #4205, #4143

This is based on 1.14.6 version. Merge ready and Gnosis merge ready if someone wants to use it.

## Changes:
- Introduces TraceStore plugin
- Allows of saving parity like traces (trace module) to database and later serving them without having the state data
- Requires block to be processed to store the trace, currently no sync options
- Node can support pruning both InMemory and Full, so while blocks need to be processed, state can be pruned.
- Supports pruning of old traces, has a limit of how many block traces needs to be kept (Head - limit ... Head], but can also keep all traces from genesis **(keep in mind that's a LOT of storage!)**
- Supports configuring what traces we want to keep
- Replaces Trace RPC module to try to grab traces from database first and fallback to state only when they are not present
- Extends Nethedmind to be able to replace JSON RPC module
- Fixes bug with `EndBlockTrace` being called too early - before applying rewards

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

I've tested by syncing chunk of Gnosis chain. And sending RPC requests. Testing further...

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs:37]: NP: Any reason why not reuse `_tracer` and remove one var?

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs:30]: There is a `0` here, is that like a RLP standard?

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs:50]: How much performance improvement when using the pool as buffer instead of passing in the gzip stream directly?

**Review by asdacap — COMMENTED** [review]
This is a very large PR, its very hard to review. Please try to separate what can be separated next time.

Overall, I don't see any problem, but I can't approve it for master without @MarekM25  or @dceleda  blessing.

**Comment by asdacap** [comment]
Also, got an estimation of storage usage?

**Comment by gituser** [comment]
> Also, got an estimation of storage usage?

I've been running this plugin both on testnet and mainnet in production.

Here is a mainnet figure for 100K blocks `TraceStore` with trace types `Trace,Rewards` plugin this configuration:

```
"TraceStore": {
    "Enabled": true, // enables
    "BlocksToKeep": 100000, // how many blocks to keep from head, default 10k, 0 - all
    "TraceTypes": "Trace,Rewards" // what traces
  }
```
```
du -sh ~/.nethermind/nethermind_db/mainnet/tracestore
3.4G	/home/nethermind/.nethermind/nethermind_db/mainnet/tracestore
```

Apart from few bugs (which Luk is aware of) plugin is very nice and saves a lot of space gives you an ability to trace transactions without running archive node.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs:52]: Placeholder for empty value, I want to have a differentiation with just empty bytes.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs]: None, no idea why I didn't do that to begin with.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs:37]: Interface mismatch, some methods are only allowed through interface

**Comment by LukaszRozmej** [comment]
The size isn't that bad and depends on blocks itself, how many of them to keep (pruning) ect. 
With moderate size pruning, keeping 10k-100k blocks, size is small (XXX MB - X GB).
Whole gnosis chain is like XX GB.

**Comment by tkstanczak** [comment]
It was supported in the past (I built it in 2018) but then I decided to remove it so simplify code.
(you can bring code back for comparison if that helps)

Are you sure you wan to add it back @LukaszRozmej?
Can I get more details on who requested it and justification for adding the feature?

**Comment by LukaszRozmej** [comment]
> It was supported in the past (I built it in 2018) but then I decided to remove it so simplify code. (you can bring code back for comparison if that helps)
> 
> Are you sure you wan to add it back @LukaszRozmej? Can I get more details on who requested it and justification for adding the feature?

Multiple people requested it, including GraphQL where it was deemed crucial to properly support Gnosis chain, as well as multiple people who wanted to migrate from OE on Gnosis chain.

File: src/Nethermind/Nethermind.Core/Collections/IListExtensions.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/Collections/IListExtensions.cs:130]: Why `- startIndex`?

File: src/Nethermind/Nethermind.Evm/Tracing/ITxTracer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.Evm/Tracing/ITxTracer.cs:406]: Are these autogenerated?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/Tracing/ITxTracer.cs:406]: No but I decided to finally document it, yes it should be in separate PR.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceSerializer.cs:72]: Why not put in test code then?

File: src/Nethermind/Nethermind.Core/Collections/IListExtensions.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Collections/IListExtensions.cs:130]: looks like a bug, but I think I refactored code to not use this method so I will just delete it for now.

**Comment by tkstanczak** [comment]
> > It was supported in the past (I built it in 2018) but then I decided to remove it so simplify code. (you can bring code back for comparison if that helps)
> > Are you sure you wan to add it back @LukaszRozmej? Can I get more details on who requested it and justification for adding the feature?
> 
> Multiple people requested it, including GraphQL where it was deemed crucial to properly support Gnosis chain, as well as multiple people who wanted to migrate from OE on Gnosis chain.
> 
> Historically we also had similar inquiries from mainnet users migrating of OE.
> 
> Main problems are depth and performance of trace module request without this storage.
> 
> This being opt-in plugin makes it abstract enough from main functionality.

Cool, thanks.
Have a look at the historical code, I think it was removed around mid 2019

**Comment by LukaszRozmej** [comment]
The observer regression was due to: #4472 reverted in #4914

**Comment by LukaszRozmej** [comment]
Gnosis chain archive traces weight currently 110GB

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs:37]: QQ: Why not pass the serializer directly?

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:204]: NP: Another idea is to create a `Map` method to `ResultWrapper`.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:230]: NP: I'm guessing returning trace directly and checking `traces is null` would also work.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:241]: NP: `index < traces.Count` is not necessary right?

**Review by asdacap — APPROVED** [review]
Don't see any significant issue.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs:37]: Old design choice, fixed.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ParityTxTraceFromStore.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ParityTxTraceFromStore.cs:54]: Yes, that is by design.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:204]: Not sure if I understood, but I think like this it is better separated to map just trace types instead of JSON RPC types

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:218]: Nested `if` makes this hard, so there are multiple conditions to return from this path.

File: src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc.TraceStore/TraceStoreRpcModule.cs:241]: right

**Comment by sambacha** [comment]
awesome

File: src/Nethermind/Nethermind.JsonRpc/Modules/RpcModuleProvider.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/RpcModuleProvider.cs:5]: super minor cosmetics - redundant new comment lines. Can be safely ignored

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ParityTxTraceFromReplay.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ParityTxTraceFromReplay.cs:5]: super minor cosmetics - redundant new comment lines. Can be safely ignored

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs:5]: super minor cosmetics - redundant new comment lines. Can be safely ignored

File: src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs:138]: super minor cosmetics - redundant comma. Can be safely ignored

**Review by marcindsobczak — APPROVED** [review]
Looks correct, I didn't find any bug

File: src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs:138]: I like those commas, if you add new line you don't have to change this one

Comment (asdacap) [reply src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs:143]: Yes, all language should allow this.

**Pull request merged**
